homeassistant/components/maico/sensor.py

```diff
@@ -0,0 +1,161 @@
+"""Sensor for retrieving Maico Centralized ventilation device information."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+# # from datetime import timedelta
+import logging
+
+from homeassistant.components.sensor import (
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+
+# # from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import entity, entity_registry
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+# # from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
+# from homeassistant.util import Throttle
+from .const import DOMAIN
+from .coordinator import MaicoUpdater
+
+LOGGER = logging.getLogger(__name__)
+
+# ICON_HAPPY = "mdi:fan-chevron-down"
+# ICON_OTHER = "mdi:fan"
+# ICON_SAD = "mdi:fan-alert"
+
+
+# class MaicoData:
+#     """Maico Data object."""
+
+#     def __init__(self, interval, maico_url):
+#         """Init Maico Data object."""
+#         self.maico_details = None
+#         self.update = Throttle(interval)(self._update)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Add sensors for passed config_entry in HA."""
+    coordinator: MaicoUpdater = hass.data[DOMAIN][config_entry.entry_id]
+    created = set()
+    all_sensors = coordinator.data["sensors"]
+
+    @callback
+    def _create_entity(key: str) -> MaicoSensor:
+        """Create a sensor entity."""
+        created.add(key)
+        # data = coordinator.data["sensors"][key]
+        # description = ENTITY_DESCRIPTION_KEY_MAP.get(
+        #     data.getUnit(), IotaWattSensorEntityDescription("base_sensor")
+        # )
+        description = MaicoSensorEntityDescription(
+            "capteur maico", state_class=SensorStateClass.MEASUREMENT
+        )
+
+        return MaicoSensor(
+            coordinator=coordinator,
+            key=key,
+            entity_description=description,
+        )
+
+    async_add_entities(_create_entity(key) for key in all_sensors)
+
+    @callback
+    def new_data_received():
+        """Check for new sensors."""
+        entities = [_create_entity(key) for key in all_sensors if key not in created]
+        async_add_entities(entities)
+
+    coordinator.async_add_listener(new_data_received)
+
+
+@dataclass
+class MaicoSensorEntityDescription(SensorEntityDescription):
+    """Class describing Maico sensor entities."""
+
+    value: Callable | None = None
+
+
+class MaicoSensor(CoordinatorEntity[MaicoUpdater], SensorEntity):
+    """Defines a IoTaWatt Energy Sensor."""
```

Comment:
IotaWatt? :)

Reply:
As it's my first integration, there too many copy paste from Iotawatts integration. Some cleaning is missing ;-)